mtenn/combination.py

```diff
@@ -0,0 +1,413 @@
+import torch
+
+
+class Combination(torch.nn.Module):
+    def forward(self, pred_list, grad_dict, param_names, *model_params):
+        """
+        This function signature should be the same for any Combination subclass
+        implementation.
+
+        Parameters
+        ----------
+        pred_list: List[torch.Tensor]
+            List of delta G predictions to be combined using LSE
+        grad_dict: dict[str, List[torch.Tensor]]
+            Dict mapping from parameter name to list of gradients
+        param_names: List[str]
+            List of parameter names
+        model_params: torch.Tensor
+            Actual parameters that we'll return the gradients for. Each param
+            should be passed individually for the backward pass to work right.
+        """
+        raise NotImplementedError("Must implement the `forward` method.")
+
+    @staticmethod
+    def split_grad_dict(grad_dict):
+        """
+        Helper method used by all Combination classes to split up the passed grad_dict
+        for saving by context manager.
+
+        Parameters
+        ----------
+        grad_dict : Dict[str, List[torch.Tensor]]
+            Dict mapping from parameter name to list of gradients
+
+        Returns
+        -------
+        List[str]
+            Key in grad_dict corresponding 1:1 with the gradients
+        List[torch.Tensor]
+            Gradients from grad_dict corresponding 1:1 with the keys
+        """
+        # Deconstruct grad_dict to be saved for backwards
+        grad_dict_keys = [
+            k for k, grad_list in grad_dict.items() for _ in range(len(grad_list))
+        ]
+        grad_dict_tensors = [
+            grad for grad_list in grad_dict.values() for grad in grad_list
+        ]
+
+        return grad_dict_keys, grad_dict_tensors
+
+    @staticmethod
+    def join_grad_dict(grad_dict_keys, grad_dict_tensors):
+        """
+        Helper method used by all Combination classes to reconstruct the grad_dict
+        from keys and grad tensors.
+
+        Parameters
+        ----------
+        grad_dict_keys : List[str]
+            Key in grad_dict corresponding 1:1 with the gradients
+        grad_dict_tensors : List[torch.Tensor]
+            Gradients from grad_dict corresponding 1:1 with the keys
+
+        Returns
+        -------
+        Dict[str, List[torch.Tensor]]
+            Dict mapping from parameter name to list of gradients
+        """
+        # Reconstruct grad_dict
+        grad_dict = {}
+        for k, grad in zip(grad_dict_keys, grad_dict_tensors):
+            try:
+                grad_dict[k].append(grad)
+            except KeyError:
+                grad_dict[k] = [grad]
+
+        return grad_dict
+
+
+class MeanCombination(Combination):
+    """
+    Combine a list of predictions by taking the mean.
+    """
+
```

Comment:
`MaxCombination` has a `__repr__` and a `__str__`, either make one here or get rid of other one (s).

Reply:
the reason I implemented it for that one is that `MaxCombination` has parameters that I wanted included in the `repr`. for the other two, the default `repr` that's implemented by `pytorch` is fine. I can still add it in though for them if you think it'd be better

Reply:
Nah thats fine as is.